lib/cn/route.c

```diff
@@ -155,22 +155,40 @@ struct route_map {
     struct route_node     rtm_nodev[] HSE_L1D_ALIGNED;
 };
 
-static int
-route_node_key_set(
-    struct route_map  *map,
-    struct route_node *node,
-    const void        *edge_key,
-    uint               edge_klen)
+static merr_t
+route_node_keybuf_alloc(struct route_node *node, uint edge_klen)
 {
-    if (!edge_key || !edge_klen)
-        return 0;
+    uint32_t kbufsz = node->rtn_keybufsz;
+
+    /* 1. If the edge key fits the inline buffer then use the inline buffer.
+     * 2. If the edge key doesn't fit the inline buffer but fits the previously allocated
+     * buffer then use the previously allocated buffer.
+     * 3. If the edge key doesn't fit both the inline and the previously allocated buffer
+     * then allocate a new larger buffer.
+     *
+     * For cases (1) and (3), free the previously allocated buffer.
+     */
+    if ((edge_klen <= sizeof(node->rtn_keybuf)) || edge_klen > kbufsz) {
+        if (kbufsz > 0) {
+            assert(node->rtn_keybufp != node->rtn_keybuf);
+            free(node->rtn_keybufp);
+            node->rtn_keybufp = node->rtn_keybuf;
+            node->rtn_keybufsz = 0;
+        }
 
-    INVARIANT(node);
-    INVARIANT(node->rtn_keybufp);
-    assert(node->rtn_keybufp != node->rtn_keybuf || edge_klen <= sizeof(node->rtn_keybuf));
+        if (edge_klen > sizeof(node->rtn_keybuf)) {
+            size_t align = sizeof(uint64_t);
 
-    memcpy(node->rtn_keybufp, edge_key, edge_klen);
-    node->rtn_keylen = edge_klen;
+            assert(edge_klen > kbufsz);
+            kbufsz = ALIGN(edge_klen, align);
+
+            node->rtn_keybufp = aligned_alloc(align, kbufsz);
+            if (!node->rtn_keybufp)
+                return merr(ENOMEM);
+
+            node->rtn_keybufsz = kbufsz;
+        }
+    }
```

Comment:
We'll need to ensure we have some tests that exercise this with a mix of short and long keys and lots of splits and joins..

Reply:
Agreed, will open a ticket to track that.